src/SegmentedArray.chpl

```diff
@@ -0,0 +1,59 @@
+module SegmentedArray {
+    use AryUtil;
+    use CTypes;
+    use MultiTypeSymbolTable;
+    use MultiTypeSymEntry;
+    use ServerConfig;
+    use Reflection;
+    use Logging;
+    use ServerErrors;
+
+    private config const logLevel = ServerConfig.logLevel;
+    const saLogger = new Logger(logLevel);
+
+    proc getSegArray(name: string, st: borrowed SymTab, type eltType): owned SegArray throws {
+        var abstractEntry = st.lookup(name);
+        if !abstractEntry.isAssignableTo(SymbolEntryType.SegArraySymEntry) {
+            var errorMsg = "Error: Unhandled SymbolEntryType %s".format(abstractEntry.entryType);
+            saLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);
+            throw new Error(errorMsg);
+        }
+        var entry: SegArraySymEntry = abstractEntry: borrowed SegArraySymEntry(eltType);
+        return new owned SegArray(name, entry, eltType);
+    }
+
+    /*
+    * This version of the getSegString method takes segments and values arrays as
+    * inputs, generates the SymEntry objects for each and passes the
+    * offset and value SymTab lookup names to the alternate init method
+    */
+    proc getSegArray(segments: [] int, values: [] ?t, st: borrowed SymTab): owned SegArray throws {
+        var segmentsEntry = new shared SymEntry(segments);
+        var valuesEntry = new shared SymEntry(values);
+        var segEntry = new shared SegArraySymEntry(segmentsEntry, valuesEntry, t);
+        var name = st.nextName();
+        st.addEntry(name, segEntry);
+        return getSegArray(name, st, segEntry.etype);
+    }
+
```

Comment:
this looks good! 

We don't need this for this PR, but we should look into making `assembleSegArrayFromParts` methods. Maybe something like
```chapel
proc assembleSegArrayFromParts(offsets:GenSymEntry, values:GenSymEntry, type etype, st:borrowed SymTab): owned SegArray throws {
    var offs = toSymEntry(offsets, int);
    var vals = toSymEntry(values, etype);
    // This probably invokes a copy, but it's a temporary legacy bridge until we can pass
    // array components as a single message.
    return getSegArray(offs.a, vals.a, st);
}

proc assembleSegArrayFromParts(offsets:SymEntry, values:SymEntry, type etype, st:borrowed SymTab): owned SegArray throws {
    var segArrayEntry = new shared SegArraySymEntry(segmentsEntry, valuesEntry, etype);
    var name = st.nextName();
    st.addEntry(name, segArrayEntry);
    return getSegArray(name, st, segEntry.etype);
}
```
There might need some typechecking boilerplate and a case for segstrings, but that's left as an exercise for the reader :)

Reply:
This is not possible. I built the function you recommend and due to the type checking (in order to create the proper SegArray), the return type is not able to be determined by the compiler and therefore it will not compile properly.